message MSG_RL_QLearning {

	// In the model only the ParameterSpaces are needed.
	// The rest (states and actions) is being inferred.
	message Model {
		message ParameterSpace {
			required string name               = 1;
			required int32  minValue           = 2;
			required int32  maxValue           = 3;
			required int32  defaultValue       = 4;
			required double impact             = 5; 
			required bool   withActions        = 6;
		}
		repeated ParameterSpace parameterSpace = 1;
	}


	message Pair {
		required int32 action   = 1;
		required float value    = 2;
	}


	message Matrix {
		message Row {
			required int32 state = 1;
			repeated Pair entry  = 2;
		}
		repeated Row row         = 2;
	}


	required Model model               = 1;
	required double learningRate       = 2;
	required double discountRate       = 3;
	required double traceDecayRate     = 4;
	required int32  traceLength        = 5;
	required int32 maxRadialDistance   = 6;
	required Matrix Q                  = 7;

}
